### ISSUE TITLE
Modify operators for simd_mask to better match std::bitset

### ISSUE THREAD

**Issue by danieltowner** [mattkretz/std-simd-feedback#30]
simd_mask is close in behaviour to std::bitset, but unlike std::bitset it doesn't provide NOT (~) or shift operators (<<, >>). I'm not sure how useful the shift operators actually are, but for completeness I think they should be provided.

simd_mask does provide && and || but I'm not sure they really make sense. These are meant to be short-circuit operators, but, like any customer operator of this type, they don't actually short-circuit. The simd_mask is a collection of bits, and so only providing the bitwise operators would seem to make more sense to me.

I think that the operator==/!= are also confusing. I would understand an expression like `my_mask == your_mask' to mean `are the masks the same', not `which elements in the masks are the same'. Looking to std::bitset for guidance again, that uses the former meaning (and only provides == anyway_ since != is synthesised), and returns a single value indicating equality for the entire value, not its elements.

**Comment by mattkretz** [comment]
* The idea is for `operator~` to be equivalent to `operator!`, right? Fine with me.
* Shift operators would have to act element-wise, otherwise it would be inconsistent. So shifting on `bool`. I'd rather not.
* `operator&&` and `||` cannot reasonably be short-circuiting even if the language allowed it (there have been proposals - EWG doesn't want it bad enough). Well yes, it could do an `all_of` / `none_of` reduction first to stop evaluation... But as I see it:

  1. To support generic code, `x < y && y > 0` needs to compile with `T` and `simd<T>`.
  2. When I need to do a reduction I either write `all_of(x < y) && all_of(y > 0)` or `all_of(x < y && y > 0)`. I don't use `&` there; that's just strange looking syntax. But if `&&` were actually short-circuiting, the second expression must do a `none_of` reduction under the hood and introduce a branch. I don't think that's helping.

* Your point on `operator==(simd_mask, simd_mask)` etc. has some history. My original proposal returned `bool`. But WG21 asked for consistency with `simd`, which is why they return a mask now. I buy the consistency argument: *every operator* acts element-wise. I think https://wg21.link/p0851 might help.

TL;DR: I think I'm okay with adding unary `~` to mean unary `!`.

**Comment by danieltowner** [comment]
Yes, unary~ is unary!

I meant shift operators on simd_mask only, not simd itself. I'd use an interface like std::rotate, std::shift_left and so on for simd itself (and in turn they can be implemented using permute). It was only a suggestion for consistency with std::bitset though which does have shift operators. I'm not that keen to have them so we could just agree to ignore this one.

Understood re: &&||. Keep them to make things work with simd or non-simd code.

I get the point about consistency with other operators. For me, the closest thing to simd_mask is std::bitset, and std::bitset's behaviour has bit-wise operators for everything except equality, so there is already precedence for making equality generate just a bool; we introduce one consistency but at the expense of another. I still think equality to bool makes more sense, but if WG21 have spoken, we don't have a choice do we?

**Comment by mattkretz** [comment]
Actually, unary `~` should *not* be unary `!`.
```c++
int x = ...;
auto y = ~(x > 0);
```
The type of `y` is `int`, not `bool`. So
```c++
simd<int> x = ....;
auto y = ~(x > 0);
```
`y` shouldn't be a `simd_mask` here. Same return type as unary `+` and `-`, I'd say.

**Comment by mattkretz** [comment]
I also believe that aiming for consistency between `simd_mask` and `bitset` is misguided. We need to aim for consistency with `bool`.
```c++
int x = ...;
auto y = (x > 0) << 1;
```
Again, the type of `y` is `int`, not `bool`.
I'm seeing a pattern here.
This could all be so much easier if EWG accepts P2600's proposal to do ADL on the return type of `operator T()`.

**Comment by mattkretz** [comment]
To elaborate. I think of `simd_mask` as a bag of `bool` values (not `bool` objects, though; i.e. no representation implied, only semantics). I think of `bitset` as an interface to interact with bits in contiguous memory. Specifically, `simd_mask` and `bitset` have different semantics. By being able to convert between them you can switch to different semantics, whenever needed.
Shifting multiple `bool`s is different from shifting contiguous bits.

Maybe even `operator[]` on `simd` and `simd_mask` is a mistake. It's the only (or am I missing another?) operator that doesn't apply element-wise. What if we ever extend `simd` to support `simd<int[4]>`? (Not that I see why; but the element type has a subscript operator that could be applied element-wise.)
